Objects/genobject.c

```diff
@@ -230,18 +234,33 @@ gen_send_ex(PyGenObject *gen, PyObject *arg, int exc, int closing)
             /* Delay exception instantiation if we can */
             if (PyAsyncGen_CheckExact(gen)) {
                 PyErr_SetNone(PyExc_StopAsyncIteration);
+                Py_CLEAR(result);
             }
             else if (arg) {
-                /* Set exception if not called by gen_iternext() */
-                PyErr_SetNone(PyExc_StopIteration);
+                if (is_return_value != NULL) {
+                    *is_return_value = 1;
+                }
+                else {
+                    /* Set exception if not called by gen_iternext() */
```

Comment:
Is not a caller responsible for raising StopIteration?

Reply:
This function is used in multiple places and some of them currently assume exception will be raised on their behalf - updating them is a possibility but will cause more churn in the code

Reply:
> This function is used in multiple places and some of them currently assume exception will be raised on their behalf 

Yes, let's not change that.